tools/ci/check_grep2.py

```diff
@@ -0,0 +1,519 @@
+
+    >>> bool(check_mixed_indentation(None, ["\\t\\t foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["\\t \\t foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["\\t // foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, [" \\tfoo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["  \\t  foo"]))
+    True
+
+    >>> bool(check_mixed_indentation(None, ["\\t  * foo"]))
+    False
+    >>> bool(check_mixed_indentation(None, ["\\t\\t* foo"]))
+    False
+    >>> bool(check_mixed_indentation(None, ["\\t \\t  * foo"]))
+    False
+    """
+    for line in lines:
+        if re.match(r"^(\t+ | +\t)\s*[^\s\*]", line):
+            return "ERROR: Mixed <tab><space> indentation detected, please stick to tab indentation."
+
+
+def check_trailing_newlines(filename, lines):
+    last_line = [x for x in lines][-1]
+    if not last_line.endswith("\n"):
+        return f"ERROR: File {filename} is missing a trailing newline"
+
+
+def check_global_vars(filename, lines):
+    for line in lines:
+        if re.match(r"^/*var/", line):
+            return "ERROR: Unmanaged global var use detected in code, please use the helpers."
+
+
+def check_proc_args_with_var_prefix(filename, lines):
+    for line in lines:
+        if re.match(r"^/[\w/]\S+\(.*(var/|, ?var/.*).*\)", line):
+            return "ERROR: Changed files contains a proc argument starting with 'var'"
+
+
+BASIC_MAP_ISSUES = OrderedDict(
+    {
+        "basic_map_issues": (
+            check_tgm,
+            check_comments,
+            check_conflict_markers,
+            check_conflict_marker_object,
+            check_iconstate_tags,
+            check_step_varedits,
+            check_pixel_varedits,
+            check_area_varedits,
+            check_base_turf_type,
+            check_common_spelling_mistakes,
+        ),
+    }
+)
+
+FULL_CONTENT_MAP_ISSUES = OrderedDict(
+    {
+        "full_content_map_issues": (
+            check_multiple_lattices,
+            check_multiple_airlocks,
+            check_multiple_firelocks,
+            check_apc_pixel_shifts,
+            check_lattice_and_wall_stacking,
+            check_window_and_wall_stacking,
+            check_airlock_and_wall_stacking,
+            check_grille_and_window_stacking,
+            check_multiple_turfs,
+            check_multiple_areas,
+        ),
+    }
+)
+
+CODE_SECTIONS = OrderedDict(
+    {
+        "515 proc_syntax": (check_515_proc_syntax,),
+        "whitespace issues": (
+            check_space_indentation,
+            check_mixed_indentation,
+            check_trailing_newlines,
+        ),
+        "common mistakes": (
+            check_global_vars,
+            check_proc_args_with_var_prefix,
+        ),
+    }
+)
+
+MAP_FILE_CACHE = {}
+
+
+if __name__ == "__main__":
+    print("check_grep2 started")
+
+    failure = False
+
+    start = time.time()
+    for section, parts in BASIC_MAP_ISSUES.items():
```

Comment:
For the map ones (if performance matters much) you'll probably be faster opening the files first and then running the checks on each one, though I appreciate that'll also cause some trouble with how it logs each test in order.

Reply:
All the map checks are out of here now, happily